### PR TITLE
Remove keyMirror in PropagationPhases

### PR DESCRIPTION
**Pull request by vjeux**
See rationale in https://www.facebook.com/groups/2003630259862046/permalink/2097088000516271/

screen shot 2016-08-28 at 10 54 34 am

### CODE REVIEW

**Comment by zpao** [comment]
I don't think we can do this one. Because the keys in the object in SimpleEventPlugin use `captured` as the key (no quotes), then *if* gcc crushes those keys then React will stop working.

**Comment by gaearon** [comment]
Good catch, thanks.

File: src/renderers/shared/stack/event/EventPropagators.js [hunk 1]
```diff
@@ -11,22 +11,22 @@
...
 var getListener = EventPluginHub.getListener;
 
 /**
  * Some event types have a notion of different registration names for different
  * "phases" of propagation. This finds listeners by a given phase.
  */
-function listenerAtPhase(inst, event, propagationPhase) {
+function listenerAtPhase(inst, event, propagationPhase: PropagationPhases) {
   var registrationName =
     event.dispatchConfig.phasedRegistrationNames[propagationPhase];
```
Comment (gaearon) [inline src/renderers/shared/stack/event/EventPropagators.js]: @vjeux 

To be specific, the problem is on this line.
GCC will crush [lines like these](https://github.com/vjeux/react/blob/02afcecd643603ec47d882424c48c9ff5663e067/src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js#L38-L43):

```js
{
  abort: {
    phasedRegistrationNames: {
      bubbled: keyOf({onAbort: true}),
      captured: keyOf({onAbortCapture: true}),
    },
  },
}
```

into

```js
{
  a: {
    b: {
      c: keyOf({e: true}),
      d: keyOf({f: true}),
    },
  },
}
```

However, `phase` will be a full string:

```js
var phase = upwards ? 'bubbled' : 'captured';
```

and so

```js
 var registrationName =
      event.dispatchConfig.phasedRegistrationNames[propagationPhase];		     
```

will give you `undefined`.

**Comment by vjeux** [comment]
After asking on twitter, no one has actually successfully used React with GCC Advanced mode itself in a while and all the ClojureScript libraries just use React as is: https://twitter.com/Vjeux/status/770338982110429186

The plan forward is to kill with fire all the abstractions that we have in place that attempts to try to support GCC Advanced mode in order to make the codebase easier to understand.

If we want to support it or some other similar optimizations going forward, we should start a separate effort for it.

**Pull request merged**